scripts/update_lit_checks.py

```diff
@@ -0,0 +1,171 @@
+#!/usr/bin/env python3
+# Copyright 2021 WebAssembly Community Group participants
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A test case update script.
+
+This script is a utility to update wasm-opt based lit tests with new FileCheck
+patterns. It is based on LLVM's update_llc_test_checks.py script.
+"""
+
+import argparse
+import glob
+import os
+import re
+import subprocess
+import sys
+import tempfile
+
+
+script_name = os.path.basename(__file__)
+NOTICE = f';; NOTE: Assertions have been generated by {script_name}'
+RUN_LINE_RE = re.compile(r'^\s*;;\s*RUN:\s*(.*)$')
+CHECK_PREFIX_RE = re.compile(r'.*--check-prefix[= ](\S+).*')
+FUNC_RE = re.compile(r'^\s*\(func \$(\S*).*$', re.MULTILINE)
+
+
+def warn(msg):
+    print(f'WARNING: {msg}', file=sys.stderr)
+
+
+def itertests(args):
+    'Yield (filename, lines) for each test specified in the command line args'
+    for pattern in args.tests:
+        tests = glob.glob(pattern)
+        if not tests:
+            warn(f'No tests matched {pattern}. Ignoring it.')
+            continue
+        for test in tests:
+            with open(test) as f:
+                lines = [line.rstrip() for line in f]
+            first_line = lines[0] if lines else ''
+            if script_name not in first_line and not args.force:
+                warn(f'Skipping test {test} which was not generated by '
+                     f'{script_name}. Use -f to override.')
+                continue
+            yield test, lines
+
+
+def find_run_lines(test, lines):
+    matches = [m.group(1)
+               for m in [RUN_LINE_RE.match(l) for l in lines]
+               if m]
```

Comment:
I would personally split such double comprehensions into two for readability, but maybe this code is coming directly from the upstream version so it makes sense to preserve it?